presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/MergeWindows.java

```diff
@@ -85,30 +85,45 @@ protected PlanNode visitPlan(
                 RewriteContext<Multimap<WindowNode.Specification, WindowNode>> context)
```

Comment:
@kokosing: AD NP-hardness of this problem.

Let’s define the directed graph of dependencies between window nodes.
Each node in such graph stands for a WindowNode.
An edge connecting nodes A and B in such graph, means that WindowNode A depends on WindowNode B.
For any possible, legal query plan below constraints are met:
1. named graph does not have to be connected (there could be WindowNodes not depending on other WindowNodes)
2. the graph is acyclic (feature of any SQL plan)

We are looking for subsets of nodes such that for each node in a subset, there’s no edge between that node and any other node in the subset, e.g. for graph V: {A,B,C} E: {(A,B)} the subsets would look like: {[A,C], [B]} or {[B,C], [A]}.
For each element in the graph, there’s a function F defined which returns the set where the element belongs.
The key here is to find the minimal number of such subsets.
Rename subset to label and there’s a definition of graph coloring (there are no two adjacent nodes sharing the same color [being in the same subset]).

This is a proof of WindowNodes dependency fulfilling -> coloring graph correspondence.
Below is for the opposite direction:

SQL plan is a tree (connected acyclic graph). Dependency between WindowNodes in legal plan can be established only from parent to child (or many children).
(induction):
f(1) - graph with single node is always convertible into correct plan (is acyclic, there is not a single node depending on parent)
f(n) -> f(n+1) - assuming acyclic graph of n nodes is linearizable into a SQL plan, for adding a new node we have a few possibilities:
1. add a disconnected node -> then we can put it in any place in the plan because it does not depend on any other node, nor any node depends on it
2. add a node which is connected with inbound edge, then put this node below the deepest located in the plan parent
3. add a node which is connected with outbound edge, then put such a node above its highest child in the plan
4. add a node connected in and out, both at the same time, then move this node above any dependent nodes and move the parent (node connected with current node using inbound edge), with all his dependencies, above that node. It’s always possible because:
 - there’s a guarantee to not have cycles in our input graph, so the added node cannot introduce a cycle
 - f(n) is already linearized (prerequisite of an implication of the induction)

—

Given the above proof in both directions: our problem -> graph coloring, graph coloring -> our problem, I claim this problem is NP-hard to solve, unless you can prove our graph has more features (e.g. coloring a bipartite graph is not a NP problem) which may help graph coloring (technically, oriented coloring because we talk about directed graphs here).